File: examples/java-junit-maven/src/test/java/examples/LibrarySteps.java
package examples;

import com.oselvar.var.Registrar;
import com.oselvar.var.State;
import com.oselvar.var.StateBinder;
import com.oselvar.var.StepDefinitions;
import java.time.LocalDate;
import java.util.List;
import java.util.Locale;
import java.util.regex.Pattern;
import java.util.stream.Stream;

public final class LibrarySteps implements StepDefinitions {

    record Ctx(List<Library.Loan> loans, int feePence, boolean granted) implements State {}

    private static final List<String> MONTHS = List.of(
            "January",
            "February",
            "March",
            "April",
            "May",
            "June",
            "July",
            "August",
            "September",
            "October",
            "November",
            "December");

    /** June 6th → LocalDate 2026-06-06 (the spec's year is 2026). */
    private static LocalDate toDate(String raw) {
        String[] parts = raw.split(" ");
        int day = Integer.parseInt(parts[1].substring(0, parts[1].length() - 2));
        return LocalDate.of(2026, MONTHS.indexOf(parts[0]) + 1, day);
    }

    /** £2.50 and 50p, both as pence. */
    private static int toPence(String raw) {
        return raw.startsWith("£")
                ? (int) Math.round(Double.parseDouble(raw.substring(1)) * 100)
                : Integer.parseInt(raw.substring(0, raw.length() - 1));
    }

    @Override
    public void defineSteps(Registrar registrar) {
        registrar.defineParameterType(
                "date",
                Pattern.compile("(?:January|February|March|April|May|June"
                        + "|July|August|September|October|November|December)"
                        + " \\d{1,2}(?:st|nd|rd|th)"),
                groups -> toDate(groups[0]));
        registrar.defineParameterType(
                "money",
                Pattern.compile("£\\d+(?:\\.\\d{2})?|\\d+p"),
                groups -> toPence(groups[0]),
                // The inverse: mismatches render as £2.60 / 50p, not a bare pence int.
                pence -> pence < 100 ? pence + "p" : String.format(Locale.ROOT, "£%.2f", pence / 100.0));
        // Emphasis (*Emma*) is stripped before matching, so a title is a
        // Title Case run in the plain prose.
        registrar.defineParameterType("title", Pattern.compile("[A-Z][a-z]+(?: [A-Z][a-z]+)*"), groups -> groups[0]);

        StateBinder<Ctx> s = registrar.defineState(() -> new Ctx(List.of(), 0, false));

        s.stimulus(
                "borrowed {title}, due back on {date}",
                (Ctx ctx, String title, LocalDate due) -> new Ctx(
                        Stream.concat(ctx.loans().stream(), Stream.of(new Library.Loan(title, due)))
                                .toList(),
                        ctx.feePence(),
                        ctx.granted()));

        s.stimulus(
                "returns it on {date}",
                (Ctx ctx, LocalDate returnedOn) -> new Ctx(
                        ctx.loans(),
                        ctx.loans().stream()
                                .mapToInt(loan -> Library.lateFee(loan, returnedOn))
                                .sum(),
                        ctx.granted()));

        s.sensor("owes a {money} late fee", (Ctx ctx, Integer expected) -> ctx.feePence());

        s.sensor("{money} for each day overdue", (Ctx ctx, Integer expected) -> Library.FEE_PENCE_PER_DAY);

        s.stimulus(
                "asks to borrow {title} on {date}",
                (Ctx ctx, String title, LocalDate on) ->
                        new Ctx(ctx.loans(), ctx.feePence(), Library.mayBorrow(ctx.loans(), on)));

        s.sensor("the library refuses", (Ctx ctx) -> {
            if (ctx.granted()) throw new AssertionError("expected the library to refuse");
            return null;
        });

        s.sensor("the library agrees", (Ctx ctx) -> {
            if (!ctx.granted()) throw new AssertionError("expected the library to agree");
            return null;
        });
    }
}

File: examples/python-pytest/steps/library.steps.py
from datetime import date

from library_example import FEE_PENCE_PER_DAY, late_fee, may_borrow
from var import define_state

MONTHS = [
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December",
]


def to_date(raw):
    """June 6th → date(2026, 6, 6) (the spec's year is 2026)."""
    month, day = raw.split(" ")
    return date(2026, MONTHS.index(month) + 1, int(day[:-2]))


def to_pence(raw):
    """£2.50 and 50p, both as pence."""
    return round(float(raw[1:]) * 100) if raw.startswith("£") else int(raw[:-1])


stimulus, sensor = define_state(
    lambda: {"loans": (), "fee_pence": 0, "granted": False},
    {
        "date": {
            "regexp": (
                r"(?:January|February|March|April|May|June"
                r"|July|August|September|October|November|December)"
                r" \d{1,2}(?:st|nd|rd|th)"
            ),
            "parse": to_date,
        },
        "money": {
            "regexp": r"£\d+(?:\.\d{2})?|\d+p",
            "parse": to_pence,
            # The inverse: mismatches render as £2.60 / 50p, not a bare pence int.
            "format": lambda pence: f"{pence}p" if pence < 100 else f"£{pence / 100:.2f}",
        },
        # Emphasis (*Emma*) is stripped before matching, so a title is a
        # Title Case run in the plain prose.
        "title": {"regexp": r"[A-Z][a-z]+(?: [A-Z][a-z]+)*"},
    },
)


@stimulus("borrowed {title}, due back on {date}")
def _(state, title, due):
    return {"loans": (*state["loans"], {"title": title, "due": due})}


@stimulus("returns it on {date}")
def _(state, returned_on):
    return {"fee_pence": sum(late_fee(loan, returned_on) for loan in state["loans"])}


@sensor("owes a {money} late fee")
def _(state, expected):
    return state["fee_pence"]


@sensor("{money} for each day overdue")
def _(state, expected):
    return FEE_PENCE_PER_DAY


@stimulus("asks to borrow {title} on {date}")
def _(state, title, on):
    return {"granted": may_borrow(state["loans"], on)}


@sensor("the library refuses")
def _(state):
    if state["granted"]:
        raise AssertionError("expected the library to refuse")


@sensor("the library agrees")
def _(state):
    if not state["granted"]:
        raise AssertionError("expected the library to agree")
